bdk_electrum_example/src/electrum.rs

```diff
@@ -0,0 +1,150 @@
+use std::{collections::BTreeMap, ops::Deref};
+
+use bdk_cli::{anyhow::Result, Broadcast};
+use bdk_core::{
+    bitcoin::{Script, Txid},
+    sparse_chain::SparseChain,
+    BlockId, TxHeight,
+};
+use electrum_client::{Client, Config, ElectrumApi};
+
+pub struct ElectrumClient {
+    client: Client,
+}
+
+impl ElectrumClient {
+    pub fn new(url: &str) -> Result<Self> {
+        let client = Client::from_config(url, Config::default())?;
+        Ok(Self { client })
+    }
+}
+
+impl Deref for ElectrumClient {
+    type Target = Client;
+    fn deref(&self) -> &Self::Target {
+        &self.client
+    }
+}
+
+impl Broadcast for ElectrumClient {
+    type Error = electrum_client::Error;
+    fn broadcast(&self, tx: &bdk_core::bitcoin::Transaction) -> Result<(), Self::Error> {
+        let _ = self.client.transaction_broadcast(tx)?;
+        Ok(())
+    }
+}
+
+/// We will detect reorg if it has depth less than this
+const REORG_DETECTION_DEPTH: u32 = 100;
+const DEFAULT_STOP_GAP: usize = 10;
+
+impl ElectrumClient {
+    /// Fetch latest block height.
+    pub fn get_height(&self) -> Result<u32> {
+        // TODO: unsubscribe when added to the client, or is there a better call to use here?
+        Ok(self
+            .client
+            .block_headers_subscribe()
+            .map(|data| data.height as u32)?)
+    }
+
+    /// Scan for a list of scripts, and create an initial [`ChainGraph`] candidate update.
+    /// This update will only contain [`Txid`]s in SparseChain, and no actual transaction data.
+    ///
+    /// User needs to fetch the required transaction data and update the [`ChainGraph`] before applying it.
+    pub fn spk_txid_scan(
+        &self,
+        spks: impl Iterator<Item = Script>,
+        last_known_height: Option<u32>,
+    ) -> Result<SparseChain> {
+        let mut dummy_keychains = BTreeMap::new();
+        dummy_keychains.insert((), spks.enumerate().map(|(i, spk)| (i as u32, spk)));
+
+        Ok(self
+            .wallet_txid_scan(dummy_keychains, None, last_known_height)?
+            .0)
+    }
+
+    /// Scan for a keychain tracker, and create an initial [`KeychainScan`] candidate update.
+    /// This update will only contain [`Txid`]s in SparseChain, and no actual transaction data.
+    ///
+    /// User needs to fetch the required transaction data and update the [`KeychainScan`] before applying it.
+    pub fn wallet_txid_scan<K: Ord + Clone>(
+        &self,
+        scripts: BTreeMap<K, impl Iterator<Item = (u32, Script)>>,
+        stop_gap: Option<usize>,
+        last_known_height: Option<u32>,
+    ) -> Result<(SparseChain, BTreeMap<K, u32>)> {
+        let mut sparse_chain = SparseChain::default();
+
+        // 1. Create checkpoint data from last_known_height - REORG_DEPTH
+        // Reorg situation will he handled at the time of applying this KeychainScan.
+        // If there's reorg deeper than the assumed depth, update process will throw error.
```

Comment:
I guess this comment is stale? I don't see an error thrown.

Reply:
The intention was to say that the error will occur while performing `apply_update` with the `KeychainScan`.. Thats what I meant by the `update process will throw error.`.. Maybe can be made more clear?

Reply:
Ah sorry I didn't read this. Let's design it so this error case doesn't exist :). I don't see how you're meant to recover from the error with this API.